x/merkledb/db.go

```diff
@@ -255,53 +251,47 @@ func newDatabase(
 	}
 
 	// mark that the db has not yet been cleanly closed
-	err = trieDB.metadataDB.Put(cleanShutdownKey, didNotHaveCleanShutdown)
+	err = trieDB.baseDB.Put(cleanShutdownKey, didNotHaveCleanShutdown)
 	return trieDB, err
 }
 
 // Deletes every intermediate node and rebuilds them by re-adding every key/value.
 // TODO: make this more efficient by only clearing out the stale portions of the trie.
 func (db *merkleDB) rebuild(ctx context.Context) error {
 	db.root = newNode(nil, RootPath)
-	if err := db.nodeDB.Delete(rootKey); err != nil {
-		return err
-	}
-	it := db.nodeDB.NewIterator()
-	defer it.Release()
 
-	viewSizeLimit := math.Max(
-		db.nodeCache.maxSize/rebuildViewSizeFractionOfCacheSize,
+	opsSizeLimit := math.Max(
+		db.valueNodeDB.nodeCache.maxSize/rebuildViewSizeFractionOfCacheSize,
 		minRebuildViewSizePerCommit,
 	)
-	currentOps := make([]database.BatchOp, 0, viewSizeLimit)
+	intermediateNodeIt := db.baseDB.NewIteratorWithPrefix(intermediateNodePrefix)
+	defer intermediateNodeIt.Release()
+	if err := database.ClearPrefix(db.baseDB, db.baseDB, intermediateNodePrefix); err != nil {
+		return err
+	}
 
-	for it.Next() {
-		if len(currentOps) >= viewSizeLimit {
+	// add all key/values back into the database
+	currentOps := make([]database.BatchOp, 0, opsSizeLimit)
+	valueIt := db.NewIterator()
+	defer valueIt.Release()
```

Comment:
This isn't a change in this PR - but we probably don't want to hold a DB iterator with all these changes happening. We should probably occasionally release and re-acquire the iterator periodically (to avoid DB size increasing). Because we are only performing DB writes... I don't 100% know that this increases the DB size a lot... but I think it can.